src/aspire/basis/fle_2d.py

```diff
@@ -0,0 +1,746 @@
+        coeffs = coeffs.copy().reshape(-1, self.count)
+        num_img = coeffs.shape[0]
+        coeffs *= self.h * self.norm_constants
+        coeffs = coeffs.T
+
+        out = np.zeros(
+            (self.num_interp, 2 * self.max_ell + 1, num_img),
+            dtype=np.float64,
+            order="F",
+        )
+        for i in range(self.ell_p_max + 1):
+            out[:, i, :] = self.A3_T[i] @ coeffs[self.idx_list[i]]
+        out = np.moveaxis(out, -1, 0)
+        if self.num_interp > self.num_radial_nodes:
+            out = dct(out, axis=1, type=2)
+            out = out[:, : self.num_radial_nodes, :]
+            out = idct(out, axis=1, type=2)
+
+        return out
+
+    def _step2(self, betas):
+        """
+        Adjoint of _step2_t and Step 2 of the forward transformation (coefficients
+            to images).
+        Uses the IFFT to convert Beta values into Fourier-space images.
+        """
+        num_img = betas.shape[0]
+        tmp = np.zeros(
+            (num_img, self.num_radial_nodes, self.num_angular_nodes),
+            dtype=np.complex128,
+        )
+
+        betas = np.swapaxes(betas, 0, 2)
+        betas = betas.reshape(-1, self.num_radial_nodes * num_img)
+        betas = self.r2c_nus @ betas
+        betas = betas.reshape(-1, self.num_radial_nodes, num_img)
+        betas = np.swapaxes(betas, 0, 2)
+
+        tmp[:, :, self.nus] = np.conj(betas)
+        z = fft.ifft(tmp, axis=2)
+
+        return z
+
+    def _step1(self, z):
+        """
+        Adjoint of _step1_t and final step of the forward transformation (coefficients
+            to images).
+        Performs the NUFFT on Fourier-space images to compute real-space images.
+        """
+        num_img = z.shape[0]
+        z = z[:, :, : self.num_angular_nodes // 2].reshape(num_img, -1)
+        im = anufft(
+            z,
+            np.stack((self.grid_x, self.grid_y)),
+            (self.nres, self.nres),
+            epsilon=self.epsilon,
+        )
+        im = im + np.conj(im)
+        im = np.real(im)
+        im = im.reshape(num_img, self.nres, self.nres)
+        im[:, self.radial_mask] = 0
+
+        return im
+
+    def create_dense_matrix(self):
+        """
+        Directly computes the transformation matrix from Cartesian coordinates to
+        FLE coordinates without any shortcuts.
+        :return: A NumPy array of size `(self.nres**2, self.count)` containing the matrix
+            entries.
+        """
+        # See Eqns. 3 and 4, Section 1.2
+        ts = np.arctan2(self.ys, self.xs)
+
+        B = np.zeros((self.nres, self.nres, self.count), dtype=np.complex128, order="F")
+        for i in range(self.count):
+            B[:, :, i] = self.basis_functions[i](self.rs, ts) * self.h
+        B = B.reshape(self.nres**2, self.count)
+        B = transform_complex_to_real(np.conj(B), self.ells)
+        B = B.reshape(self.nres**2, self.count)
+        if self.match_fb:
+            B[:, self.flip_sign_indices] *= -1.0
```

Comment:
Relative to the FB ordering!

Reply:
Sorry, yes, it would be the other way around. I double checked the `evaluate_t` and `evaluate` too and you're right.  So I had it backwards and it's relative to FLE ordering. Thanks!